### PR TITLE
Bjorncs/zstd java

### PR DESCRIPTION
**Pull request by bjorncs**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: vespajlib/src/test/java/com/yahoo/compress/CompressorTest.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        byte[] inputData = "The quick brown fox jumps over the lazy dog".getBytes();
+        Compressor compressor = new Compressor(CompressionType.ZSTD);
+        Compressor.Compression compression = compressor.compress(CompressionType.ZSTD, inputData, Optional.empty());
+        assertEquals(inputData.length, compression.uncompressedSize());
+        byte[] compressedData = compression.data();
+        byte[] decompressedData = compressor.decompress(CompressionType.ZSTD, compressedData, 0, inputData.length, Optional.of(compressedData.length));
+        assertArrayEquals(inputData, decompressedData);
+    }
+
+}
```
Comment (baldersheim) [inline vespajlib/src/test/java/com/yahoo/compress/CompressorTest.java:27]: missing newline

File: vespajlib/src/test/java/com/yahoo/compress/ZstdCompressorTest.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    @Test
+    void compresses_and_decompresses_input() {
+        byte[] inputData = "The quick brown fox jumps over the lazy dog".getBytes();
+        ZstdCompressor compressor = new ZstdCompressor();
+        byte[] compressedData = compressor.compress(inputData, 0, inputData.length);
+        byte[] decompressedData = compressor.decompress(compressedData, 0, compressedData.length);
+        assertArrayEquals(inputData, decompressedData);
+    }
+
+}
```
Comment (baldersheim) [inline vespajlib/src/test/java/com/yahoo/compress/ZstdCompressorTest.java:22]: missing newline

File: vespajlib/src/test/java/com/yahoo/compress/ZstdOuputStreamTest.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            zstdOut.write(inputData[0]);
+            zstdOut.write(inputData, 1, inputData.length - 1);
+        }
+        byte[] compressedData = arrayOut.toByteArray();
+        ZstdCompressor compressor = new ZstdCompressor();
+        byte[] decompressedData = new byte[inputData.length];
+        compressor.decompress(compressedData, 0, compressedData.length, decompressedData, 0, decompressedData.length);
+        assertArrayEquals(inputData, decompressedData);
+    }
+}
```
Comment (baldersheim) [inline vespajlib/src/test/java/com/yahoo/compress/ZstdOuputStreamTest.java:30]: missing newline

File: vespajlib/src/test/java/com/yahoo/compress/ZstdOuputStreamTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    @Test
+    void output_stream_compresses_input() throws IOException {
+        byte[] inputData = "The quick brown fox jumps over the lazy dog".getBytes();
+        ByteArrayOutputStream arrayOut = new ByteArrayOutputStream();
+        try (ZstdOuputStream zstdOut = new ZstdOuputStream(arrayOut, 12)) {
+            zstdOut.write(inputData[0]);
+            zstdOut.write(inputData, 1, inputData.length - 1);
+        }
+        byte[] compressedData = arrayOut.toByteArray();
+        ZstdCompressor compressor = new ZstdCompressor();
```
Comment (baldersheim) [inline vespajlib/src/test/java/com/yahoo/compress/ZstdOuputStreamTest.java:25]: You should check that the data has actually been compressed. Simplest is just to check that the output is shorter than the input.

File: vespajlib/src/test/java/com/yahoo/compress/ZstdCompressorTest.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * @author bjorncs
+ */
+class ZstdCompressorTest {
+
+    @Test
+    void compresses_and_decompresses_input() {
+        byte[] inputData = "The quick brown fox jumps over the lazy dog".getBytes();
+        ZstdCompressor compressor = new ZstdCompressor();
+        byte[] compressedData = compressor.compress(inputData, 0, inputData.length);
+        byte[] decompressedData = compressor.decompress(compressedData, 0, compressedData.length);
```
Comment (baldersheim) [inline vespajlib/src/test/java/com/yahoo/compress/ZstdCompressorTest.java:18]: You should check that the data has actually been compressed. Simplest is just to check that the output is shorter than the input.

**Review by baldersheim — CHANGES_REQUESTED** [review]

File: vespajlib/src/test/java/com/yahoo/compress/ZstdOuputStreamTest.java (hunk 4 above)
Comment (bjorncs) [reply vespajlib/src/test/java/com/yahoo/compress/ZstdOuputStreamTest.java:26]: I have added a new test case for that. The compressed size is actually larger for this test string.

**Comment by bjorncs** [comment]
@baldersheim PTAL

**Review by baldersheim — APPROVED** [review]

**Pull request merged**